shopinvader/services/cart.py

```diff
@@ -68,7 +68,85 @@ def clear(self):
         cart = self._clear_cart(cart)
         return self._to_json(cart)
 
+    def _get_validator_cart_by_id_domain(self, value):
+        """
+        Get cart by id domain. Limiting to session partner.
+        :param value:
+        :return:
+        """
+        return [("partner_id", "=", self.partner.id), ("id", "=", value)]
+
+    def _get_line_copy_vals(self, line):
+        """
+        Prepare copy values to be passed to _add_item
+        :param line: sale.order.line
+        :return: dict
+        """
+        return {
+            "product_id": line.product_id.id,
+            "item_qty": line.product_uom_qty,
+        }
+
+    def _get_lines_to_copy(self, cart):
+        return cart.order_line
+
+    # pylint: disable=W8102,W8106
+    def copy(self, **params):
+        """
+        This service allows
+        :param params:
+        :return:
+        """
+        return self._copy(**params)
+
+    def _copy(self, **params):
+        """
+        Copy the cart given by id without the lines
+        They will be re-added
+        :return: dict/json
+        """
+        cart = self.env["sale.order"].search(
+            self._get_validator_cart_by_id_domain(params.get("id"))
+        )
+        # Copy the existing cart
```

Comment:
We have two issue

1: copying a SO will keep the typology to "sale" indead of "cart"
2: copying a SO with special product (delivery, product discount, ...) will add then as normal product

Maybe it's simplier to use odoo to copy the line and then call the _shopinvader_recompute https://github.com/shopinvader/odoo-shopinvader/blob/a056abc9b94f5919149ee515ddb140779efa5022/shopinvader/models/sale.py#L70

Reply:
Yes, but that method didn't exist at that time :smile:

Reply:
> 1: copying a SO will keep the typology to "sale" indead of "cart"

We can override default value for that.

Reply:
> 2: copying a SO with special product (delivery, product discount, ...) will add then as normal product

There is a _get_lines_to_copy() method for that purpose:

https://github.com/shopinvader/odoo-shopinvader/pull/558/files#diff-c22517a22e5a715e4f408262c310f0cdR90

Reply:
As this behaviour is included in shopinvader, we can propagate that mechanism to other modules if needed.

Reply:
> > 1: copying a SO will keep the typology to "sale" indead of "cart"
> 
> We can override default value for that.
I think it's better to keep "sale" as default value as I a user create a SO in odoo he expect that the SO is in the typo "sale"


> As this behaviour is included in shopinvader, we can propagate that mechanism to other modules if needed.

Yes indeed, maybe we should do it in this PR?

Reply:
> I think it's better to keep "sale" as default value as I a user create a SO in odoo he expect that the SO is in the typo "sale"
> 
> >

I wanted to say in the copy() call in our code.
https://github.com/shopinvader/odoo-shopinvader/pull/558/files#diff-c22517a22e5a715e4f408262c310f0cdR113

Reply:
> > I think it's better to keep "sale" as default value as I a user create a SO in odoo he expect that the SO is in the typo "sale"
> > >
> 
> I wanted to say in the copy() call in our code.
> https://github.com/shopinvader/odoo-shopinvader/pull/558/files#diff-c22517a22e5a715e4f408262c310f0cdR113

Yes here ! ;)

Reply:
So, there, I can add default value : {'typology': 'cart'}

Reply:
yes

Reply:
@sebastienbeau Typology is "cart" in copy()

Reply:
> 2: copying a SO with special product (delivery, product discount, ...) will add then as normal product

For delivery, the code is included in shopinvader_delivery_carrier(in this PR)